data/src/history/manager.rs

```diff
@@ -439,13 +439,53 @@ impl Data {
 
         let filtered = messages
             .iter()
-            .filter(|message| {
+            .enumerate()
+            .filter(|(message_index, message)| {
                 if let message::Source::Server(Some(source)) = message.target.source() {
-                    !server_messages.get(source).exclude
+                    match server_messages.get(source).exclude {
+                        Exclude::All => false,
+                        Exclude::None => true,
+                        Exclude::Smart(seconds) => 'search_previous_messages: {
+                            for previous_message in
+                                messages.iter().rev().skip(messages.len() - *message_index)
+                            {
+                                let duration_seconds = message
+                                    .server_time
+                                    .signed_duration_since(previous_message.server_time)
+                                    .num_seconds();
+
+                                if duration_seconds <= seconds {
+                                    if let message::Source::User(previous_message_user) =
+                                        previous_message.target.source()
+                                    {
+                                        let substring = match source {
+                                            message::source::Server::Join => {
+                                                format!("⟶ {}", previous_message_user.nickname())
+                                            }
+                                            message::source::Server::Part => {
+                                                format!("⟵ {}", previous_message_user.nickname())
+                                            }
+                                            message::source::Server::Quit => {
+                                                format!("⟵ {}", previous_message_user.nickname())
+                                            }
+                                        };
+
+                                        if message.text.starts_with(&substring) {
+                                            break 'search_previous_messages true;
```

Comment:
I don't love that we need to match on message text here because we don't persist the user information associated to the server join / part / quit messages. 

We should revisit message::Source and see if we can add user somewhere so we have strong data to check against

Reply:
Thank you for weighing in on this.  I agree entirely.  I will plow forward with my attempt to alter message::Source then, after addressing other issues with this PR :)